### ISSUE TITLE
Configure systemd-networkd to leave weave bridge, veth interfaces unmanaged (also avoid undesired DHCP or IPv6 autoconf behavior)

### ISSUE THREAD

**Issue by SpComb** [kontena/kontena#1264]
CoreOS uses systemd-networkd for network configuration, and the default CoreOS configuration will attempt DHCP + IPv6 autoconf/DHCPv6 configuration on all weave bridge/openvswitch/dummy interfaces.

The systemd 229 netlink operations also appear to interfere with the weaver netlink code: https://github.com/weaveworks/weave/issues/2479#issuecomment-257377740

We should disable DHCP and IPv6 autoconf on all of the interfaces used by weave:

* `datapath` (Type=ether Driver=openvswitch)
* `weave` (Type=ether Driver=bridge)
* `vxlan-6784` (Type=ether Driver=vxlan)
* `dummy0` (Type=ether Driver=dummy)

CoreOS already includes configuration for the Driver=veth interfaces that disables DHCP, although it does not disable IPv6 autoconf:

```
[Match]
Driver=veth

[Network]
```

This would probably be something like:

```
[Match]
Name=weave datapath vxlan-* dummy*
Driver=bridge openvswitch vxlan dummy

[Network]
DHCP=no
IPv6AcceptRA=false
```

#### `cat /usr/lib64/systemd/network/zz-default.network`
```
[Network]
DHCP=yes

[DHCP]
UseMTU=true
UseDomains=true
```

#### `$ networkctl `
```
IDX LINK             TYPE               OPERATIONAL SETUP     
  1 lo               loopback           carrier     configured
  2 eth0             ether              routable    configured
  3 eth1             ether              routable    configured
  4 docker_gwbridge  ether              no-carrier  configuring
  5 docker0          ether              routable    configured
  6 datapath         ether              degraded    configuring
  8 weave            ether              degraded    configuring
  9 vxlan-6784       ether              degraded    configuring
 10 dummy0           ether              degraded    configuring
 12 vethwe-datapath  ether              degraded    configured
 13 vethwe-bridge    ether              degraded    configured
 15 veth9f23e35      ether              degraded    configured

12 links listed.
```

#### `networkctl status weave`
```  
● 8: weave
       Link File: /usr/lib64/systemd/network/99-default.link
    Network File: /usr/lib64/systemd/network/zz-default.network
            Type: ether
           State: degraded (configuring)
          Driver: bridge
      HW Address: 3a:be:cf:0f:d6:73
         Address: fe80::38be:cfff:fe0f:d673
```

#### `tcpdump -ne -i weave`
```
11:39:33.101253 c2:79:51:44:e3:87 > 33:33:00:00:00:02, ethertype IPv6 (0x86dd), length 70: fe80::c079:51ff:fe44:e387 > ff02::2: ICMP6, router solicitation, length 16
11:39:33.113300 3a:be:cf:0f:d6:73 > 33:33:00:00:00:02, ethertype IPv6 (0x86dd), length 70: fe80::38be:cfff:fe0f:d673 > ff02::2: ICMP6, router solicitation, length 16
11:39:33.114311 3a:be:cf:0f:d6:73 > ff:ff:ff:ff:ff:ff, ethertype IPv4 (0x0800), length 338: 0.0.0.0.68 > 255.255.255.255.67: BOOTP/DHCP, Request from 3a:be:cf:0f:d6:73, length 296
11:39:33.114486 3a:be:cf:0f:d6:73 > 33:33:00:01:00:02, ethertype IPv6 (0x86dd), length 102: fe80::38be:cfff:fe0f:d673.546 > ff02::1:2.547: dhcp6 inf-req
11:39:33.120209 da:fe:a9:1e:24:97 > 33:33:00:00:00:02, ethertype IPv6 (0x86dd), length 70: fe80::d8fe:a9ff:fe1e:2497 > ff02::2: ICMP6, router solicitation, length 16
11:39:33.120413 da:fe:a9:1e:24:97 > ff:ff:ff:ff:ff:ff, ethertype IPv4 (0x0800), length 338: 0.0.0.0.68 > 255.255.255.255.67: BOOTP/DHCP, Request from da:fe:a9:1e:24:97, length 296
11:39:33.120604 da:fe:a9:1e:24:97 > 33:33:00:01:00:02, ethertype IPv6 (0x86dd), length 102: fe80::d8fe:a9ff:fe1e:2497.546 > ff02::1:2.547: dhcp6 inf-req
```

**Comment by SpComb** [comment]
It seems that having systemd-networkd touch the `veth` interfaces at all will tickle netlink races between Docker/weave and systemd-networkd: https://github.com/coreos/bugs/issues/1554

This affects CoreOS stable 1185.3.0, which ships with a Docker/libnetwork workaround (https://github.com/coreos/docker/pull/29). However, this can lead to broken weave network connectivity: https://github.com/kontena/kontena/issues/1285

We can presumably avoid these issues by reconfiguring systemd-networkd to ignore the weave (bridge, veth) interfaces: https://github.com/kontena/kontena/issues/1285#issuecomment-258905209

This appears to require overriding the default CoreOS systemd-networkd configuration to explicitly only manage the desired `eth*` interface etc, because systemd-networkd does not support explicitly marking the `veth` interfaces as unconfigured: https://github.com/systemd/systemd/pull/4228

#### `/etc/systemd/network/zz-default.network`

Override the default network rules to only match the `etc*` interfaces in use on the platform:

```
[Match]
Name=eth*
```

This should be sufficient to leave the `weave` unmanaged.

#### `/etc/systemd/network/50-docker-veth.network`

Symlink -> `/dev/null` to mask the broken CoreOS attempt at leaving  `veth` interfaces unmanaged.

**Comment by SpComb** [comment]
See https://github.com/kontena/kontena-plugin-digitalocean/pull/15 for specific discussion of CoreOS systemd-networkd configuration workarounds for DigitalOcean CoreOS, in particular:

> Specifically for DO, the CoreOS zz-default.network unit seems to be unused, because the DO CoreOS ships with a /usr/share/oem/cloud-config.yml that handles the DO network interface configuration via .network units under /run/systemd/network/.

**Comment by SpComb** [comment]
For DigitalOcean specifically, the following seems like a good CoreOS cloud-config, where the exact contents of the `zz-default.network` don't really matter, as DigitalOcean  uses the CoreOS OEM cloud-config to configure the network.  The `[Match]` part will need adjustment for other platforms, to match the external/internal network interfaces provided by the platform.

```yaml
coreos:
  units:
    - name: 50-docker.network
      mask: true
    - name: 50-docker-veth.network
      mask: true
    - name: zz-default.network
      runtime: false
      content: |
        # default should not match virtual Docker/weave bridge/veth network interfaces
        [Match]
        Name=eth*

        [Network]
        DHCP=yes

        [DHCP]
        UseMTU=true
        UseDomains=true
```

Resulting systemd-networkd configuration and state:

#### ` ls -la /etc/systemd/network/`
```
total 32
drwxr-xr-x. 2 root root 4096 Nov  9 10:22 .
drwxr-xr-x. 6 root root 4096 Nov  9 10:22 ..
lrwxrwxrwx. 1 root root    9 Nov  9 10:22 50-docker-veth.network -> /dev/null
lrwxrwxrwx. 1 root root    9 Nov  9 10:22 50-docker.network -> /dev/null
-rw-r--r--. 1 root root  126 Nov  9 10:22 zz-default.network
```

#### `$ networkctl`
```
IDX LINK             TYPE               OPERATIONAL SETUP     
  1 lo               loopback           carrier     unmanaged 
  2 eth0             ether              routable    configured
  3 eth1             ether              routable    configured
  4 docker0          ether              no-carrier  unmanaged 
  5 datapath         ether              degraded    unmanaged 
  7 weave            ether              routable    unmanaged 
  8 dummy0           ether              off         unmanaged 
 10 vethwe-datapath  ether              degraded    unmanaged 
 11 vethwe-bridge    ether              degraded    unmanaged 
 12 vxlan-6784       ether              degraded    unmanaged 

10 links listed.
```

#### `$ networkctl status eth0`
```
● 2: eth0
       Link File: /usr/lib64/systemd/network/98-virtio.link
    Network File: /run/systemd/network/00-ca:27:61:83:3d:82.network
            Type: ether
           State: routable (configured)
            Path: virtio-pci-0000:00:03.0
          Driver: virtio_net
          Vendor: Red Hat, Inc
           Model: Virtio network device
      HW Address: ca:27:61:83:3d:82
         Address: 10.18.0.8
                  188.166.14.243
                  fe80::c827:61ff:fe83:3d82
         Gateway: 188.166.0.1 (ICANN, IANA Department)
             DNS: 8.8.8.8
                  8.8.4.4
```

#### `$ networkctl status eth1`
```
● 3: eth1
       Link File: /usr/lib64/systemd/network/98-virtio.link
    Network File: /run/systemd/network/00-d6:0c:db:20:8e:36.network
            Type: ether
           State: routable (configured)
            Path: virtio-pci-0000:00:04.0
          Driver: virtio_net
          Vendor: Red Hat, Inc
           Model: Virtio network device
      HW Address: d6:0c:db:20:8e:36
         Address: 10.133.1.161
                  fe80::d40c:dbff:fe20:8e36
             DNS: 8.8.8.8
                  8.8.4.4
```

**Comment by SpComb** [comment]
I speculate that the following trick *might* also work for the `zz-default.network` unit to only match "pyhsical" network interfaces:

```
[Match]
Path=*
```

**Comment by SpComb** [comment]
Tracking kontena plugin fixes to the cloud-config:

* https://github.com/kontena/kontena-plugin-packet/pull/12 https://github.com/kontena/kontena-plugin-packet/releases/tag/v0.2.2
* https://github.com/kontena/kontena-plugin-aws/pull/15 https://github.com/kontena/kontena-plugin-aws/releases/tag/v0.2.1
* https://github.com/kontena/kontena-plugin-upcloud/pull/6 https://github.com/kontena/kontena-plugin-upcloud/releases/tag/v0.2.1
* https://github.com/kontena/kontena-plugin-digitalocean/pull/15 https://github.com/kontena/kontena-plugin-digitalocean/releases/tag/v0.2.1
* https://github.com/kontena/kontena-plugin-vagrant/issues/24
* https://github.com/kontena/kontena-plugin-azure/issues/8

**Comment by dsmiley** [comment]
I'm looking forward to `kontena grid cloud-config` with the proper fix.

**Comment by jnummelin** [comment]
@dsmiley Might be actually pretty hard to create universally applicable fix in there since the proper systemd-networkd config is highly environment specific as you can see from the listed PR's for different cloud providers.

**Comment by dsmiley** [comment]
Ouch; this would mean the death of `kontena grid cloud-config` then; no?

**Comment by SpComb** [comment]
I suspect that the `[Match] Path=*` trick would work in the vast majority of cases. Most of the cloud provider platforms with CoreOS support don't even use the problematic CoreOS `zz-default.network` unit in question, they have their own OEM cloud-config that generates per-interface network units. I doubt that many cloud platforms even use DHCP in the first place?

However, the trouble is that a bad cloud-config would only be apparent after a reboot, when the CoreOS node fails to come back up, and it becomes pretty much impossible to access it anymore...

However, the current `kontena grid cloud-config` [template](https://github.com/kontena/kontena/blob/master/cli/lib/kontena/machine/cloud_config/cloudinit.yml) could definitely still be improved. Theoretically we could even instrument the agent to pick up the information required for generating the correct `[Match]` section, although that's a thing best handled by the platform OEM...

**Comment by SpComb** [comment]
CoreOS is introducing a fix for the network configuration that they ship in https://github.com/coreos/coreos-overlay/pull/2300

Overriding the `zz-default.network` to limit the match is still useful to avoid things like DHCP configuration of the weave interfaces, unless we also explicitly blacklist those instead.

**Comment by jnummelin** [comment]
Closing as this is now fixed in all the plugins.

**Comment by SpComb** [comment]
Still applies to

* https://github.com/kontena/kontena-plugin-azure/blob/master/lib/kontena/machine/azure/cloudinit.yml
* https://github.com/kontena/kontena/blob/master/cli/lib/kontena/machine/cloud_config/cloudinit.yml

**Comment by jnummelin** [comment]
I'll leave it open then. :) My bad.

**Comment by SpComb** [comment]
Closing as all plugins and `kontena grid cloud-config` will be fixed for Kontena 1.0.5
